### PR TITLE
dot for alloys, shield for protected plants

### PR DESCRIPTION
**Pull request by vsrisuknimit**
image

image

Dots to indicate whether steel and Ti has bonus from alloys (Advanced Alloys, Mercurian Alloys, Regoplastics). It does not care about Phobolog (user can already see the corp) or Unity (the policy affects everyone).

Shield to indicate whether plants are protected.

All work done by @nwai90 . I am just porting this over from his experimental branch.

### CODE REVIEW

**Comment by vsrisuknimit** [comment]
This can close #1391 .

**Comment by vsrisuknimit** [comment]
Tagging @chosta for review.

**Comment by kberg** [comment]
I love it, I really do. Two thoughts: 1) consider a modest + instead of dots and 2) the Moon expansion will provide similar protection to steel and titanium. I'd love to be able to fit in there.

I guess a third thought? Can that shield be reused for plants, animals and microbes?

(By modest, I mean something that isn't very thin, even if it's closer to a square than a plus? Like 
image


(Like the swiss flag's cross.)

**Comment by kevinb9n** [comment]
Hello! Just want to say in a friendly way that I believe this decision is a mistake, which makes the interface unfriendly to novice players of the game.

You're asking for mental gymnastics to be done, which experienced players like us have no trouble with. I personally really don't want to see the app go in that direction.

A number in the interface should have a clear and simple meaning. The clearest and simplest, by far, is to portray the value that each unit of the resource has. Yes, I am fully aware that there are disadvantages to having to show yet another numeral near all the existing ones, or to show a number of dots which can in the rarest of circumstances get as high as 7. But I feel that one must be over-weighting those disadvantages by a wide margin in order to feel that the current solution is acceptable.

If this goes through, I'm not saying it will be a disaster, only that to *me* it will be the clearest design mistake I've seen in this particular app, that's all.

**Comment by vsrisuknimit** [comment]
@kberg 

image

(1) An example of what + looks like. We could put it up for a vote. (2) There are very few cards that take away Ti or Steel. Don't see a need for a shield there. It adds too much cluster to the overview tab for too little benefit. (3) Yes, the shield is in asset and can be reused for protected animals. That would be a different PR though.

**Comment by chosta** [comment]
I'm not sure about this exact solution although it is needed.
Let me gather my thoughts in the morning and with the help of some coffeee I'll give my suggestions. I am not against it but I have concerns similar to @kevinb9n. I know this has been long overdue but lets do it the right way. We will have an indicator one way or another.

**Comment by chosta** [comment]
Just to be clear, the shield for plants works pretty well. The dot / plus is what concerns me.

**Comment by kberg** [comment]
> (2) There are very few cards that take away Ti or Steel. Don't see a need for a shield there.

Right, I agree -- what I'm saying, though, is that there's this card coming:

image

along with others that do. I'd like for there to be room.

**Comment by vsrisuknimit** [comment]
> You're asking for mental gymnastics to be done, which experienced players like us have no trouble with. I personally really don't want to see the app go in that direction.

There is no mental gymnastic requires. There should be no calculation at all. The point of the alloy indicator is NOT tell exactly how much Ti and steel cost. The point of the indicator is to tell who benefit more from Steel and Ti than other players, so that they can attack those players or hate-draft accordingly.

**Comment by vsrisuknimit** [comment]
> along with others that do. I'd like for there to be room.

Yes, there should be room for a shield. Whether we need it or not is up to debate. There are many more asteroid / plant-burning cards than Ti/Steel decreasing cards.

**Comment by kberg** [comment]
> > along with others that do. I'd like for there to be room.
> 
> Yes, there should be room for a shield. Whether we need it or not is up to debate. There are many more asteroid / plant-burning cards than Ti/Steel decreasing cards.

That's fine. When I resolve this ehnahcement with the Moon, I might have to do something such as move the shield up, to the right, and reduced size. And that's fine - we can improve from there.

**Comment by kberg** [comment]
> > You're asking for mental gymnastics to be done, which experienced players like us have no trouble with. I personally really don't want to see the app go in that direction.
> 
> There is no mental gymnastic requires. There should be no calculation at all. The point of the alloy indicator is NOT tell exactly how much Ti and steel are worth. The point of the indicator is to tell who benefit more from Steel and Ti than other players, so that they can attack those players or hate-draft accordingly.
> 
> The overview table is such a central element to the game. It should be as clean as possible.

"mental gymnastics" is a charged phrase meant to imply unfairness, when in fact this feature, as-is, already makes calculating costs easier for a player than the current state, which is to have to look up cards.

IMO, this PR, as a net improvement, is worth submitting, and iterating upon.

File: src/server/ServerModel.ts [hunk 1]
```diff
@@ -375,6 +376,7 @@ function getPlayers(players: Array<Player>, game: Game): Array<PlayerModel> {
       name: player.name,
       plants: player.plants,
       plantProduction: player.getProduction(Resources.PLANTS),
+      plantsAreProtected: player.plantsAreProtected(),
```
Comment (kberg) [inline src/server/ServerModel.ts:379]: @nwai90 here's a good use for the general purpose object representing resources. Let's expand on that idea a moment, and put a type on it:

```
interface Stuff<T> {
  plants: T;
  heat: T;
  titanium: T;
};

etc.

Then ServerModel can deliver

units: Stuff<Number>;
protected: Stuff<Boolean>

**Review by kberg — APPROVED** [review]

**Comment by vsrisuknimit** [comment]
Let's not merge until @chosta got time to sleep and have coffee.

**Comment by chosta** [comment]
I guess this is going through.
@vsrisuknimit can you check what the maximum possible amount of Ti or Steel discounts per resource? Is there going to be enough space for all of them. Also, should we show a dot / plus when unity is in charge?

**Comment by chosta** [comment]
> Let's not merge until @chosta got time to sleep and have coffee.

Oh it could be merged. I don't see any harm. If I don't like something or I have a better suggestion, I'll add it in a PR.

**Comment by vsrisuknimit** [comment]
@chosta oops. We commented at the same time. 

As I mentioned above, it only counts alloys. The point of the dot is not to help calculate the exact worth of Steel or Ti. It is to help other players know who care more about steel and Ti. So no dot for Unity, since everyone benefits. Not Phobolog since you can already read the corp name. The maximum number is 2 for both metals.

**Comment by chosta** [comment]
> As I mentioned above, it only counts alloys.

Hmm, in this case we need a legend. This got me confused as I was thinking about discounts. I still need my coffee, let's continue tomorrow.

**Comment by vsrisuknimit** [comment]
> This got me confused as I was thinking about discounts.

@chosta There is no discount here. You are getting confused with space discount from card like Space Station and Quantum Extractor.

**Comment by chosta** [comment]
> @chosta There is no discount here.

I understood that when you told me. So that is the issue here. It's not self-explanatory and raises more questions than it answers.

**Comment by vsrisuknimit** [comment]
> If it's alloys only, players need to be able to check what the symbols stand for easily.

Tooltip? On hover? But that is not good for mobile right?

**Comment by kberg** [comment]
Consider replacing dots with a MC icon. eg ([2], [3], [4]) or
[2]+1(subscript)

>

**Comment by chosta** [comment]
I think just a popup [legend] 
Could be next to the Concise / Full buttons.

We can let this pass and I'll improve once I start working on. #2150

**Comment by vsrisuknimit** [comment]
> Consider replacing dots with a MC icon. eg ([2], [3], [4]) or [2]+1(subscript)
> […](#)

@kberg That might be too distracting. You will have to show 2 and 3 for everyone otherwise it can be confusing why some players say 4 or 5.

**Comment by vsrisuknimit** [comment]
Tagging @ssimeonoff for this topic.

**Comment by kevinb9n** [comment]
Sorry, I try to avoid using charged language, so I want to revise that now. I just meant that it takes more mental effort, which I think is just factual.

Let me keep thinking about how a general-purpose Effects Tracker might possibly work, which would have the room (somehow!) to show the full values, because if it's at all possible it would take some of the pressure off here. That is, then it's easier (at least for me) to understand how this feature really only needs to serve the sort of "which player cares most about this resource" use case.

And under that model, sure, okay, it can leave Unity out, and it can skip showing the first 2/3 representing the base value. But my lingering question is: really, why not include Phobolog? Yes, you *can* see the corp name there, and *most* players know what that name means once they've played several games. But it's still going to look just plain wrong when I play Phobolog and you play Advanced Alloys and yet somehow our titanium displays are different. Also, when I see a Phobolog opponent and I see a dot there, I'm going to assume it IS the Phobolog dot and I'm not going to guess that it really means you *also* played Mercurian.

**Comment by vsrisuknimit** [comment]
> And under that model, sure, okay, it can leave Unity out, and it can skip showing the first 2/3 representing the base value. But my lingering question is: really, why not include Phobolog? Yes, you _can_ see the corp name there, and _most_ players know what that name means once they've played several games. But it's still going to look just plain wrong when I play Phobolog and you play Advanced Alloys and yet somehow our titanium displays are different. Also, when I see a Phobolog opponent and I see a dot there, I'm going to assume it IS the Phobolog dot and I'm not going to guess that it really means you _also_ played Mercurian.

I am open to adding it for Phobolog. So up to 3 dots for Ti.

**Comment by kevinb9n** [comment]
> > Consider replacing dots with a MC icon. eg ([2], [3], [4]) or [2]+1(subscript)
> > […](#)
> 
> @kberg That might be too distracting. You will have to show 2 and 3 for everyone otherwise it can be confusing why some players say 4 or 5.

(I would agree it should do it consistently like that. Note that for new players even just the 2s and 3s are a helpful reminder. But in the heat column, only Helion would have a 1, because in that case the rest are zero.)

**Comment by vsrisuknimit** [comment]
image

Some changes:
- simplify the vue methods.
- vertical arrangement instead of horizontal
- allow a dot from Phobolog. So max two dots for steels, three for Ti. Still no dot for Unity.

**Comment by vsrisuknimit** [comment]
Do not merge yet. Ongoing discussion on discord.

**Comment by kberg** [comment]
> Do not merge yet. Ongoing discussion on discord.

Sure thing. I'll leave it to other people to merge.

**Comment by bafolts** [comment]
I was confused as to what the dots represent and I am still confused as to what they represent while reading through this so the user interface may not be intuitive. The shield is intuitive to me and I assume it means you can not take those resources. The dots are the confusing part. I would still be fine merging this as we could always change the dots to something else later if while playing players arrived upon a better idea once seeing it in practice.

**Comment by bafolts** [comment]
@vsrisuknimit I think the shield seems reasonable enough. Could consider a separate PR for the protected plants while figuring out strategy for the dots.

**Comment by vsrisuknimit** [comment]
New commit allows for Unity as well.

@bafolts the dot represent metal bonus. One dot for each extra value for the metal. If you play Advanced Alloys, you get one dot each for Steel and Ti. Phobolog would have one dot for Ti from the beginning.

This allow players to tell who benefit more from the metal in case you want to attack someone's Ti or hate draft space cards from someone.

**Comment by vsrisuknimit** [comment]
image

@bafolts in this case, we can quickly tell that Green's steel is worth one more than normal while Yellow (Phobolog)'s Ti is worth 3 extra than normal.

Ready to merge I believe unless someone has a strong objection. Allowing Phobolog and Unity now after the poll result on discord. Plus symbol is not as popular as the dot. A fair number of votes for number representing the worth of the metal (lower than dots), but I dislike the clutter so I am not going to do it. 

image

**Comment by kevinb9n** [comment]
> but I dislike the clutter so I am not going to do it.

Respect!

Thanks so much for being openminded to the concerns and for taking the questions to discord. And thanks for tackling this in the first place and seeing it through. This is a great improvement.

**Comment by kberg** [comment]
I should add: there's a card in The Moon called Lunar Steel which grants an additional steel megacredit bonus.

**Comment by chosta** [comment]
We have to add a legend for these symbols in an upcoming PR but I see this as an important improvement for intermediate to experienced players. We will take care of the novice soon enough.

**Pull request merged**